### PR TITLE
Allow to pass request options (for CORS scenarios) #76

### PR DESCRIPTION
**Pull request by jackkk94**
feat: allow to pass request options

### CODE REVIEW

File: src/generators/angular/AngularServicesGenerator.ts [hunk 1]
```diff
@@ -83,25 +89,44 @@ export class AngularServicesGenerator {
...
+                            {
+                                name: 'options',
+                                place: ParameterPlace.Body,
+                                optional: true,
+                                dtoType: 'IAngularHttpRequestOptions',
+                                isCollection: false,
+                                isModel: false
+                            }
+                        );
+                    });
```
Comment (sevaru) [inline src/generators/angular/AngularServicesGenerator.ts:111]: I think this logic belongs to method generation

**Review by sevaru — CHANGES_REQUESTED** [review]

File: src/bin.ts [hunk 2]
```diff
@@ -25,6 +26,7 @@ program
     .option('--output <string>')
     .option('--configOutput <string>')
     .option('--all')
+    .option('--withRequestOptions')
```
Comment (aleshchev) [inline src/bin.ts:29]: You need describe it in Readme

File: libs/download.service.ts [hunk 3]
```diff
@@ -20,12 +20,13 @@ export class DownloadFileService extends BaseHttpService {
         url: string,
         method: 'post' | 'get',
         data?: {},
-        saveAs?: string
+        options?: IAngularHttpRequestOptions,
```
Comment (aleshchev) [inline libs/download.service.ts:23]: is must be the latest

File: src/generators/angular/AngularServicesGenerator.ts [hunk 4]
```diff
@@ -54,7 +57,10 @@ export class AngularServicesGenerator {
             {
                 kind: StructureKind.ImportDeclaration,
                 moduleSpecifier: './base-http.service',
-                namedImports: [{ name: BASE_SERVICE }]
+                namedImports: [this.settings.withRequestOptions
+                    ? { name: BASE_SERVICE }
+                    : { name: BASE_SERVICE }, { name: HTTP_REQUEST_OPTIONS }
```
Comment (aleshchev) [inline src/generators/angular/AngularServicesGenerator.ts:62]: incorrect

File: src/generators/angular/AngularServicesGenerator.ts [hunk 5]
```diff
@@ -22,15 +25,15 @@ const HTTP_CLIENT_VARIABLE_NAME = 'http';
...
+    constructor(protected aliasResolver: AliasResolver, uriBuilder: UriBuilder, private settings: IOptions) {
+        this.methodGenerator = new AngularServicesMethodGenerator(uriBuilder, this.settings);
     }
     public getServicesCodeStructure(services: IServiceModel[]): StatementStructures[] {
         return [...this.getImports(), ...this.getServices(services)];
     }
 
     private getImports(): ImportDeclarationStructure[] {
-        return [
+        const imports: ImportDeclarationStructure[] = [
```
Comment (aleshchev) [inline src/generators/angular/AngularServicesGenerator.ts:36]: why???

File: src/generators/angular/AngularServicesMethodGenerator.ts [hunk 6]
```diff
@@ -5,14 +5,15 @@ import { MethodOperation } from '../../models/kinds/MethodOperation';
...
+import { IMethodModel, IMethodParameter } from '../../models/method-parameter/IMethodModel';
 import { IPathParameter } from '../../models/method-parameter/IPathParameter';
 import { IQueryParameter } from '../../models/method-parameter/IQueryParameter';
 import { IReturnType } from '../../models/method-parameter/IReturnType';
+import { IOptions } from '../../options';
 import { UriBuilder } from '../../services/UriBuilder';
-import { first } from '../../utils';
 import { MAPPERS_NAMESPACE, MODELS_NAMESPACE, UNDEFINED_STRING } from '../utils/consts';
 import { TypeSerializer } from '../utils/TypeSerializer';
+import { HTTP_REQUEST_OPTIONS } from './AngularServicesGenerator';
```
Comment (aleshchev) [inline src/generators/angular/AngularServicesMethodGenerator.ts:17]: declare HTTP_REQUEST_OPTIONS here

File: src/generators/angular/AngularServicesMethodGenerator.ts [hunk 7]
```diff
@@ -81,6 +82,21 @@ export class AngularServicesMethodGenerator {
         return statement;
     }
 
+    protected getMethodParameters(methodModel: IMethodModel): IMethodParameter[] {
```
Comment (aleshchev) [inline src/generators/angular/AngularServicesMethodGenerator.ts:85]: wrong place for it

File: src/generators/angular/AngularServicesMethodGenerator.ts [hunk 8]
```diff
@@ -141,18 +157,20 @@ export class AngularServicesMethodGenerator {
             : `Observable<${this.getReturnTypeName(x.returnType, x.returnType?.type.type)}>`;
     }
     private createDownloadMethod(writer: CodeBlockWriter, model: IMethodModel): void {
-        const parameter = first(model.parameters.filter((z) => z.name !== 'saveAs' && z.place === ParameterPlace.Body));
+        const bodyParameters = model.parameters.filter((z) => z.name !== 'saveAs' && z.place === ParameterPlace.Body);
+        const dataParameter = bodyParameters.find(z => z.name === 'data');
+        const optionsParameter = bodyParameters.find(z => z.name === 'options');
 
-        if (!model.parameters.find((z) => z.name === 'saveAs')) {
+        if (!model.parameters.find(z => z.name === 'saveAs')) {
```
Comment (aleshchev) [inline src/generators/angular/AngularServicesMethodGenerator.ts:164]: revert (). it's style

**Review by aleshchev — CHANGES_REQUESTED** [review]

File: src/generators/angular/AngularServicesGenerator.ts (hunk 4 above)
Comment (jackkk94) [reply src/generators/angular/AngularServicesGenerator.ts:61]: why?

File: src/generators/angular/AngularServicesMethodGenerator.ts [hunk 9]
```diff
@@ -22,14 +24,14 @@ interface IParameterType {
...
-    constructor(private uriBuilder: UriBuilder) {}
+    constructor(private uriBuilder: UriBuilder, private settings: IOptions) {}
 
     public getMethodsCodeStructures(methodModels: IMethodModel[]): OptionalKind<MethodDeclarationStructure>[] {
         return methodModels.map(
             (methodModel): OptionalKind<MethodDeclarationStructure> => ({
                 scope: Scope.Public,
                 name: methodModel.name,
-                parameters: methodModel.parameters.map((p) => this.getParameterStatement(p)),
+                parameters: this.getMethodParameters(methodModel).map((p) => this.getParameterStatement(p)),
```
Comment (sevaru) [inline src/generators/angular/AngularServicesMethodGenerator.ts:34]: I guess we should separate our parameters from swagger and specific generation parameters (e.g. request options)

So I suggest solution like:

1) Method getMethodParameters(methodModel) that would return ParameterDeclarationStructure[]
2) getMethodParameters would map our methodModel.parameters
and adds option parameter as needed
Something like
```ts
public getMethodParameters(model) {
     const statements = model.parameters.map(x => this.getParameterStatement(x));

     if (this.options.withRequestOptions) {
          statements.push(this.getRequestOptionsParameter());
     }

     return statements;
}
```

that way we would not depend on the structure of our swagger based parameters.
and would not mutate model.parameters (which is external data for this service)

File: src/models/method-parameter/IMethodModel.ts [hunk 10]
```diff
@@ -5,11 +5,13 @@ import { IPathParameter } from './IPathParameter';
 import { IQueryParameter } from './IQueryParameter';
 import { IReturnType } from './IReturnType';
 
+export type IMethodParameter = IPathParameter | IQueryParameter | IBodyParameter;
```
Comment (sevaru) [inline src/models/method-parameter/IMethodModel.ts:8]: I think we should not name types with "I" and it should be MethodParameter

**Review by sevaru — CHANGES_REQUESTED** [review]

File: src/generators/angular/AngularServicesMethodGenerator.ts [hunk 11]
```diff
@@ -31,32 +32,42 @@ export class AngularServicesMethodGenerator {
...
+
+    private getRequestOptionsParameter(): OptionalKind<ParameterDeclarationStructure> {
+        return this.getParameterStatement({
+            name: 'options',
+            place: ParameterPlace.Body,
+            optional: true,
+            dtoType: HTTP_REQUEST_OPTIONS,
+            isCollection: false,
+            isModel: false
+        });
```
Comment (sevaru) [inline src/generators/angular/AngularServicesMethodGenerator.ts:70]: Maybe we could generate ParameterDeclarationStructure with method getParameterStatement?

There are alot of useless information like place, dtoType, isCollection, isModel

File: src/generators/angular/AngularServicesMethodGenerator.ts [hunk 12]
```diff
@@ -157,10 +168,8 @@ export class AngularServicesMethodGenerator {
             ? `Promise<${x.returnType?.type.type}>`
             : `Observable<${this.getReturnTypeName(x.returnType, x.returnType?.type.type)}>`;
     }
-    private createDownloadMethod(writer: CodeBlockWriter, model: IMethodModel): void {
-        const bodyParameters = model.parameters.filter((z) => z.name !== 'saveAs' && z.place === ParameterPlace.Body);
-        const dataParameter = bodyParameters.find((z) => z.name === 'data');
-        const optionsParameter = bodyParameters.find((z) => z.name === 'options');
+    private createDownloadMethod(writer: CodeBlockWriter, model: IMethodModel, options?: Partial<MethodParameter>): void {
+        const parameter = first(model.parameters.filter((z) => z.name !== 'saveAs' && z.place === ParameterPlace.Body));
```
Comment (sevaru) [inline src/generators/angular/AngularServicesMethodGenerator.ts:172]: we do not use anything from MethodParameter except for the name
maybe type should be optionsParameter?: { name: string }

File: src/generators/angular/AngularServicesMethodGenerator.ts [hunk 13]
```diff
@@ -60,14 +60,12 @@ export class AngularServicesMethodGenerator {
...
-        return this.getParameterStatement({
+        return {
             name: 'options',
-            place: ParameterPlace.Body,
-            optional: true,
-            dtoType: HTTP_REQUEST_OPTIONS,
-            isCollection: false,
-            isModel: false
-        });
+            type: '$types.TypeOrUndefined<IAngularHttpRequestOptions>',
```
Comment (sevaru) [inline src/generators/angular/AngularServicesMethodGenerator.ts:65]: Lets make it just HTTP_REQUEST_OPTIONS

**Review by sevaru — APPROVED** [review]

**Review by aleshchev — APPROVED** [review]

**Pull request merged**